FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/FhirToDataLakeProcessingJobSpliter.cs

```diff
@@ -0,0 +1,291 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using EnsureThat;
+using Microsoft.Extensions.Logging;
+using Microsoft.Health.Fhir.Synapse.Common.Logging;
+using Microsoft.Health.Fhir.Synapse.Core.Exceptions;
+using Microsoft.Health.Fhir.Synapse.Core.Extensions;
+using Microsoft.Health.Fhir.Synapse.Core.Fhir;
+using Microsoft.Health.Fhir.Synapse.Core.Jobs.Models;
+using Microsoft.Health.Fhir.Synapse.DataClient;
+using Microsoft.Health.Fhir.Synapse.DataClient.Api.Fhir;
+using Microsoft.Health.Fhir.Synapse.DataClient.Extensions;
+using Microsoft.Health.Fhir.Synapse.DataClient.Models.FhirApiOption;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Health.Fhir.Synapse.Core.Jobs
+{
+    public class FhirToDataLakeProcessingJobSpliter
+    {
+        private readonly IApiDataClient _dataClient;
+        private readonly IDiagnosticLogger _diagnosticLogger;
+        private readonly ILogger<FhirToDataLakeProcessingJobSpliter> _logger;
+
+        public FhirToDataLakeProcessingJobSpliter(
+            IApiDataClient dataClient,
+            IDiagnosticLogger diagnosticLogger,
+            ILogger<FhirToDataLakeProcessingJobSpliter> logger)
+        {
+            _dataClient = EnsureArg.IsNotNull(dataClient, nameof(dataClient));
+            _diagnosticLogger = EnsureArg.IsNotNull(diagnosticLogger, nameof(diagnosticLogger));
+            _logger = EnsureArg.IsNotNull(logger, nameof(logger));
+        }
+
+        public int LowBoundOfProcessingJobResourceCount { get; set; } = JobConfigurationConstants.LowBoundOfProcessingJobResourceCount;
+
+        public int HighBoundOfProcessingJobResourceCount { get; set; } = JobConfigurationConstants.HighBoundOfProcessingJobResourceCount;
+
+        public async IAsyncEnumerable<SubJobInfo> SplitJobAsync(string resourceType, DateTimeOffset? startTime, DateTimeOffset endTime, [EnumeratorCancellation] CancellationToken cancellationToken)
+        {
+            var totalCount = await GetResourceCountAsync(resourceType, startTime, endTime, cancellationToken);
+
+            if (totalCount < HighBoundOfProcessingJobResourceCount)
```

Comment:
seems we can also return if totalCount equals HighBoundOfProcessingJobResourceCount?